Add technical interviewing workshop (#319)
CodeYourFuture/CYF-Workshops: technical-interviews/README.md

# Technical interviews

## Learning objectives

```objectives
- [ ] Perform a technical interview.
- [ ] Explain what you're thinking when solving a problem.
- [ ] Write code in front of someone in an interview setting.
```

You will be applying for jobs, many of which require doing a technical interview.

> You should already have [read about technical interviews](https://sdc.codeyourfuture.io/guides/employability/interview/technical-interview-prep/). In this workshop, you'll be practising interviewing and being interviewed about technical content.

Reviewer: Have them think back to their Piscine interview.

## Interview questions

* [Is it a Set?](https://www.codewars.com/kata/5fddfbe6a65636001cc4fcd2)
* [Look and say numbers](https://www.codewars.com/kata/53ea07c9247bc3fcaa00084d)
* [Mix fruit juice](https://www.codewars.com/kata/5905871c00881d0e85000015)
* [Multiples of 3 or 5](https://www.codewars.com/kata/514b92a657cdc65150000006)
* [FizzBuzz Backwards](https://www.codewars.com/kata/59ad13d5589d2a1d84000020)

## Preparing to interview someone

Think about your goals as an interviewer. You want to:
* See the best someone can do - you're not trying to trick them.
* Work out how they approach problems:
  * Do they make a plan?
  * Do they think about edge cases?
  * How do they handle not knowing how to do something?
  * How easy are they to help?
* Can they solve the problem?
* Can they reason about their code?
* Do they run through their code with examples to test it?

To do this you will need to:
* Already know a good solution to the problem (ideally multiple solutions!)
* Prepare how you're going to explain the problem
* Listen to the candidate as they're being interviewed

>   * Work out when they're stuck (but give them enough time to come up with an answer before intervening)

Reviewer: Waiting long enough is something we all struggle with - mentors too - suggest here to count slowly to ten in your head

>   * If they're not telling you what they're thinking, encourage them to talk aloud

Reviewer: Again, give specific prompts. Encourage them to talk aloud. Say things like: "Talk through what you are thinking" and "Can you unpack this for me a bit more."

  * Give nudges (ideally by asking questions), not answers, to help them get unstuck
* Understand their code, and be able to work out whether it works
* Identify edge cases their code does not work for

## Giving feedback

After you interview someone, you will give them feedback. Think about what feedback you would appreciate. Some examples of topics:

> * How clearly did they communicate their ideas?

Reviewer: Connect this back to the demos we keep making them do.

* Did you always know what they were thinking/doing?
* How easy was their code to read and understand?
* How open were they to help?
* Did they come across as a person you'd want to work with?
* How well do you think they solved the problem? (Note: This doesn't mean "did they solve it the same way you would", though that can be an interesting comparison).

## Doing the interview (30 minutes per interview)

In pairs, take turns interviewing your partner. In each interview, one of you will be the interviewer and the other the candidate.

You should:
* Candidate: Open VS Code on your laptop.

> * Interviewer: Spend 2 minutes introducing the problem. It's ok to leave some details out (hopefully the candidate will ask about them), but the question should make sense.

Reviewer: Set a timer or show a timer on your screen.

* Candidate: Spend 20 minutes solving the problem. This should be an interactive dialogue, where the interviewer and candidate are talking throughout.
* Spend 8 minutes giving feedback and discussing the interview and experience.

Then swap.

## Discussion (30 minutes)

> As a whole class, discuss your experience. What went well? What did/didn't you understand? What things felt uncomfortable? What could you have done differently? Try not to focus on the technical specifics of the interview question you did - try to focus on general things that will help you in your next interview.

Reviewer: ```suggestion
As a whole class, discuss your experience. What went well? What did/didn't you understand? What things felt uncomfortable? What could you have done differently? Do not focus on the technical specifics of the interview question you did. Focus on general things that will help you in your next interview.
```
